### PR TITLE
🐛  Use Kubernetes 1.24 in quickstart and CAPD, bump to kind v0.14

### LINKED ISSUE #4621: Use systemd cgroupDriver in CAPD
**User Story**

As a  operator I would like to use the (kubeadm) default cgroupDriver (which is systemd) with CAPD.

**Detailed Description**

Starting with kubeadm 1.21 the default cgroupDriver is systemd. With the upcoming kind version we have to pin CAPD to cgroupDriver cgroupsfs, because it won't work otherwise.

More context:
* PR to pin cgroupDriver to cgroupfs: https://github.com/kubernetes-sigs/cluster-api/pull/4614
* Issue to upgrade to kind v0.11: https://github.com/kubernetes-sigs/cluster-api/issues/4503
* Slack discussion: https://kubernetes.slack.com/archives/C8TSNPY4T/p1620923791417000
* kind systemd issue: https://github.com/kubernetes-sigs/kind/issues/1726

**Anything else you would like to add:**

Goal is essentially to revert https://github.com/kubernetes-sigs/cluster-api/pull/4614 once https://github.com/kubernetes-sigs/kind/issues/1726 has been implemented and is usable for us.

/kind feature

### LINKED ISSUE #6509: Fix CAPD quick-start and E2E to adapt with changes in Kind v0.13
**What steps did you take and what happened:**
Kind v0.13 switched images to systemd cgroup driver and thus we have to adapt quick-start and E2E to work on 1.24.

In order to make this happen, 

- We have to upgrade the kind version used in our CI to 0.13
- We are going to switch the CAPD quick-start to use ClusterClass by default, so we can leverage version-aware patch in order to have a single template that works both with 1.24 as well as older version of Kubernetes
- For the same reason we are going to upgrade our E2E tests to use ClusterClass

**Anything else you would like to add:**
Users are required to user kind 0.13 to spin up management clusters on 1.24

/kind bug

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

* Use Kubernetes 1.24 in quickstart and CAPD
* Drops the non-ClusterClass based CAPD cluster template
* Bumps kind to v0.14 for v1.24 support as kind management cluster and for self-build images >= v1.24
* Migrates CI upgrade tests to use cluster class and version aware patches for cgroupDriver
* Implement workaround for DockerMachinePools which are not supported by ClusterClass to use correct cgroupDriver on upgrade tests.
* Enables `cgroupDriver: systemd` by removal of pinning to `cgroupDriver: cgroupfs`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6509
Fixes #4621

Still WIP:
- [x] check which other tests need to get migrated to topology
- [x] get all tests green
- [x] cleanup code

### CODE REVIEW

**Comment by chrischdi** [comment]
pull-cluster-api-e2e-workload-upgrade-1-23-latest-main currently fails because it tries to pull the kind image `kindest/node:v1.24.1-rc.0.10+8b1b4db3834ddf` which seems not to be a supported tag. 

```
{"ts":1652511846453.4377,"caller":"controller/controller.go:317","msg":"controller/dockermachine: Reconciler error","reconciler group":"infrastructure.cluster.x-k8s.io","reconciler kind":"DockerMachine","name":"k8s-upgrade-and-conformance-x4acz0-control-plane-sls88-wf7s7","namespace":"k8s-upgrade-and-conformance-suiysm","err":"failed to create worker DockerMachine: failure pulling container image: invalid reference format","errVerbose":"failure pulling container image: invalid reference format\nsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/docker.(*Machine).Create\n\tsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/docker/machine.go:224\nsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).reconcileNormal\n\tsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:236\nsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).Reconcile\n\tsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:159\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227\nruntime.goexit\n\truntime/asm_amd64.s:1581\nfailed to create worker DockerMachine\nsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).reconcileNormal\n\tsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:237\nsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).Reconcile\n\tsigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:159\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227\nruntime.goexit\n\truntime/asm_amd64.s:1581"}
```

I'm curious if this worked before :-)

**Comment by chrischdi** [comment]
> pull-cluster-api-e2e-workload-upgrade-1-23-latest-main currently fails because it tries to pull the kind image `kindest/node:v1.24.1-rc.0.10+8b1b4db3834ddf` which does not exist.
> 
> ```json
> {
>   "ts": 1652511846453.4377,
>   "caller": "controller/controller.go:317",
>   "msg": "controller/dockermachine: Reconciler error",
>   "reconciler group": "infrastructure.cluster.x-k8s.io",
>   "reconciler kind": "DockerMachine",
>   "name": "k8s-upgrade-and-conformance-x4acz0-control-plane-sls88-wf7s7",
>   "namespace": "k8s-upgrade-and-conformance-suiysm",
>   "err": "failed to create worker DockerMachine: failure pulling container image: invalid reference format",
>   "errVerbose": "failure pulling container image: invalid reference format
>   sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/docker.(*Machine).Create
>     sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/docker/machine.go:224
>   sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).reconcileNormal
>     sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:236
>   sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).Reconcile
>     sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:159
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227
>   runtime.goexit
>     runtime/asm_amd64.s:1581
>   failed to create worker DockerMachine
>   sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).reconcileNormal
>     sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:237
>   sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers.(*DockerMachineReconciler).Reconcile
>     sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/controllers/dockermachine_controller.go:159
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266
>   sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
>     sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227
>   runtime.goexit
>     runtime/asm_amd64.s:1581"
> }
> ```
> 
> I'm curious if this worked before :-)

Reason: I switched to clusterClass which has a patch to always inject the kind image instead of just using the default one.

Adjusted the patch so it is only enabled on stable release version tags, not the ones which e.g. look like: `v1.24.1-rc.0.10+8b1b4db3834ddf`, by checking if a `+` is inside the version name

**Comment by chrischdi** [comment]
/test pull-cluster-api-apidiff-main
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-23-latest-main

Changed last commit to replace `+` in `customImage` by `_` like the ci script does while preparing the images https://github.com/kubernetes-sigs/cluster-api/blob/main/scripts/ci-e2e-lib.sh#L219

**Comment by chrischdi** [comment]
After tests got green :tada: : did some cleanup and squashed the commits


/test pull-cluster-api-apidiff-main
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-23-latest-main

File: docs/book/src/user/quick-start.md [hunk 1]
```diff
@@ -285,10 +285,13 @@ The Docker provider is not designed for production use and is intended for devel
 
 </aside>
 
-The Docker provider does not require additional prerequisites.
-You can run:
+The Docker provider does requires to enable the `ClusterTopology` feature to support cluster upgrades from kubernetes <= v1.23 to v1.24.
```
Comment (CecileRobertMichon) [inline docs/book/src/user/quick-start.md:288]: why does Docker provider require enabling an experimental feature to upgrade to the latest k8s version? Will this apply to other providers as well or is this just for Docker? What about MachinePools (ClusterClass doesn't currently support MachinePool)?

Comment (chrischdi) [reply docs/book/src/user/quick-start.md:288]: This only applies to the docker provider / CAPD.

The TLDR is: currently we hardcode `cgroupDriver: cgroupfs` in all templates for CAPD but this does not work anymore for kubernetes >= v1.24 because of a breaking change in kind https://github.com/kubernetes-sigs/kind/releases/tag/v0.13.0 :

> * systemd cgroups driver will be used for Kubernetes v1.24.0+ (rather than 1.21.0+ when https://github.com/kubernetes/kubernetes/pull/99471, which we previously overrode).
>   * NOTE: You must use kind v0.13.0+ with Kubernetes v1.24.0+ images, and if you built your own Kubernetes v1.24.0+ image with a previous kind version you will need to re-built when switching to kind v0.13.0+.
>   * NOTE: You do not need to be using systemd on the host machine, systemd is used inside the kind node containers. We are now using it for Kubernetes pods in addition to running kubelet, containerd etc.
>   * KIND will continue to use cgroupfs for Kubernetes versions prior to v1.24.0.

Some history:

* kubeadm started to default to `cgroupDriver: systemd` for Kubernetes >= v1.21 (which is what the other providers also use afaik)
* Kind and its images which are used for CAPD did not support `cgroupDriver: systemd` for Kubernetes <= v1.23
* Because of that all templates for CAPD did hardcode `cgroupDriver: cgroupfs`, but beginning with v1.24 this does not work anymore.
* Therefor Kubernetes >= v1.24 requires us to not set `cgroupDriver: cgroupfs` anymore.

Perspective for the Docker/CAPD quickstart guide:

* If somebody uses the quickstart guide and provisions a cluster having version v1.23
* To upgrade the cluster to v1.24 the person would have to:
  * Remove the `cgroupDriver: cgroupfs` entries at the KCP
  * Change the kubernetes version at the KCP
  * Create a new `KubeadmConfigTemplate` for the `MachineDeployment`
  * Change the `MachineDeployment` to refer the new `KubeadmConfigtemplate`
  * Change the kubernetes version at the `MachineDeployment`

With ClusterClass and the version-aware patch for `cgroupDriver`: a user would only need to change the kubernetes version at the Cluster object.

Regarding MachinePools in Docker/CAPD:

* When a `MachinePool` gets updates from v1.23 -> v1.24: the same steps as above for `MachineDeployment` are required (new `KubeadmConfigTemplate` and changing `MachinePool`)
* As soon as #5991 is implemented, an version-aware patch could also get added for `MachinePools`

Perspective for the E2E tests:

* Its way easier and requires less temporary hacks (until Kubernetes <= v1.23 does not get tested anymore) to use ClusterClass and the version aware patch
* Only CAPD tests using MachinePools require the two temporary hacks which are in this PR.

I hope this brings some light in the dark and helps to continue the discussion.

Some more context is at #6509 :-)

If we decide to continue as is I could of course add some more information at the quickstart guide about why it is required to enable ClusterTopology for docker/CAPD.

File: test/framework/machinepool_helpers.go [hunk 2]
```diff
@@ -139,6 +141,21 @@ func UpgradeMachinePoolAndWait(ctx context.Context, input UpgradeMachinePoolAndW
...
+
+		// This is a workaround for CAPD which has to use cgroupDriver: cgroupfs for kubernetes clusters < v1.24.
+		// It is required because kind and the node images do not support the default systemd for kubernetes < v1.24.
+		// This is to be removed as soon as no upgrades from kubernetes < v1.24 is tested anymore with CAPD.
+		if mp.Spec.Template.Spec.InfrastructureRef.Kind == "DockerMachinePool" {
+			if semver.MustParse(strings.TrimPrefix(input.UpgradeVersion, "v")).GTE(semver.MustParse("1.24.0")) {
+				// Ensure that we update to a BootstrapConfig which does not use cgroupfs
+				mp.Spec.Template.Spec.Bootstrap.ConfigRef.Name = strings.TrimSuffix(mp.Spec.Template.Spec.Bootstrap.ConfigRef.Name, "-cgroupfs")
+				// Reset DataSecretName so a new one gets created.
+				if strings.HasSuffix(*mp.Spec.Template.Spec.Bootstrap.DataSecretName, "-cgroupfs") {
```
Comment (sbueringer) [inline test/framework/machinepool_helpers.go:153]: If I understand this correctly just changing the bootstrap ref in a MP will not be rolled out? If that's the case we should probably create an issue for it. I might be wrong but this sounds like a bug to me

**Review by sbueringer — COMMENTED** [review]
First round of review

File: docs/book/src/user/quick-start.md [hunk 3]
```diff
@@ -732,19 +735,11 @@ The Docker provider is not designed for production use and is intended for devel
 
 </aside>
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply docs/book/src/user/quick-start.md:737]: A compromise overall could be to say: if you need upgrades to work use development-topology otherwise use development

[the remark this answers is not in the document]
Comment (chrischdi) [reply docs/book/src/user/quick-start.md:737]: The compromise seems fair to me

* having the `development` flavor as second option
* add the note

[the remark this answers is not in the document]
Comment (killianmuldoon) [reply docs/book/src/user/quick-start.md:737]: >if you need upgrades to work use development-topology otherwise use development

That seems confusing from my POV. Maybe we could make the topology version the default version 'development' and change the current non-ClusterClass to 'deployment-legacy', leave it in the guide and add a note about upgrades not working correctly with the legacy version.

File: test/e2e/cluster_upgrade_test.go [hunk 4]
```diff
@@ -67,7 +83,7 @@ var _ = Describe("When upgrading a workload cluster with a single control plane
 			SkipConformanceTests:     true,
 			ControlPlaneMachineCount: pointer.Int64(1),
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:84]: Yes, except the WorkerMachineCount wich defaults to `2` at the other test.

Should we remove one of them?

File: test/e2e/cluster_upgrade_test.go [hunk 5]
```diff
@@ -20,14 +20,30 @@ limitations under the License.
...
-	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
 )
 
 var _ = Describe("When upgrading a workload cluster and testing K8S conformance [Conformance] [K8s-Upgrade]", func() {
 	ClusterUpgradeConformanceSpec(ctx, func() ClusterUpgradeConformanceSpecInput {
+
+		// For KubernetesVersionUpgradeFrom < v1.24 we have to use upgrades-cgroupfs flavor as source cluster.
+		// This is because kind and CAPD require to use cgroupDriver cgroupfs for Kubernetes <v1.24
+		// and systemd for >=v1.24.
+		if version := e2eConfig.Variables[KubernetesVersionUpgradeFrom]; semver.MustParse(strings.TrimPrefix(version, "v")).LT(semver.MustParse("1.24.0")) {
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go]: migrated to ParseTolerant, but ignoring the error. We can't return an error here. WDYT?

File: test/framework/machinepool_helpers.go (hunk 2 above)
Comment (chrischdi) [reply test/framework/machinepool_helpers.go:153]: I'll recheck that

Comment (chrischdi) [reply test/framework/machinepool_helpers.go:153]: My bad, looks like that was another issue while testing. I'll remove that part.

**Comment by chrischdi** [comment]
I hope I was able to address all currently solvable findings correctly. 😃 

Open discussion points from my side / which are not completely clear to me:

* The upgrade tests for `When upgrading a workload cluster using ClusterClass` and `When upgrading a workload cluster with a single control plane machine` are the same except `WorkerMachineCount` (2 vs 1). Should we drop one of these?
  * xref https://github.com/kubernetes-sigs/cluster-api/pull/6513#discussion_r876684441
 
* Discussion about `development` and `development-topology` flavor, current outlined options:
  * Keep both flavors and add a note to the non-ClusterClass flavor variant due to upgrade problematic, swap their order to make topology the first option:
    * rename both, e.g.: `development` -> `development-legacy` && `development-topology` -> `development`
    * only rename `development` -> `development-legacy`
    * keep names
  * Remove `development` flavor
    * Keep `development-topology` as-is
    * rename `development-topology` to `development`
  * What is the best here? I would currently favor Keeping both flavors, adding a note to the non-ClusterClass flavor, swap their order to make topology the first option and at least rename `development` -> `development-legacy`

  * xref https://github.com/kubernetes-sigs/cluster-api/pull/6513#discussion_r874839209
 
* semver parsing and ignoring the error
  * Should we ignore the error? In worst case this leads to a broken upgrade which would also have been the case prior to this change.
  * xref https://github.com/kubernetes-sigs/cluster-api/pull/6513#discussion_r874844201

File: test/e2e/cluster_upgrade_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go]: Sidenote: ignoring the error leads to the same result as this code would have without this changes.

File: test/framework/machinepool_helpers.go (hunk 2 above)
Comment (chrischdi) [reply test/framework/machinepool_helpers.go]: Seems like I have to take another look, it breaks for the `pull-cluster-api-e2e-workload-upgrade-1-23-latest-main` test

File: hack/ensure-kind.sh [hunk 6]
```diff
@@ -21,7 +21,7 @@ set -o pipefail
 set -x
 
 GOPATH_BIN="$(go env GOPATH)/bin/"
-MINIMUM_KIND_VERSION=v0.12.0
+MINIMUM_KIND_VERSION=v0.13.0
```
Comment (sbueringer) [inline hack/ensure-kind.sh:24]: Note to myself: bump to v0.14.0

**Comment by sbueringer** [comment]
Current state:
* I have to update the user-facing templates based on our slack discussion
* I reintroduced setting the MP dataSecretName to nil during upgrade as it's required to trigger a rollout of the new bootstrapconfig
* let's wait for green tests

/test pull-cluster-api-e2e-workload-upgrade-1-23-latest-main
/test pull-cluster-api-e2e-full-main

File: docs/book/src/user/quick-start.md (hunk 1 above)
Comment (sbueringer) [reply docs/book/src/user/quick-start.md]: Created Slack thread for it: https://kubernetes.slack.com/archives/C8TSNPY4T/p1653416265784229

Comment (sbueringer) [reply docs/book/src/user/quick-start.md]: Based on the thread I only kept the ClusterClass template. Please let me know if there are any objections (@CecileRobertMichon @fabriziopandini)

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-workload-upgrade-1-23-latest-main
/test pull-cluster-api-e2e-full-main

@fabriziopandini Ready for review / merge in my opinion.
I think this PR is cherry-pickable as is for v1.24 support in v1.1.x

/cherry-pick release-1.1

File: test/infrastructure/docker/examples/machine-pool.yaml [hunk 7]
```diff
@@ -22,14 +22,20 @@ spec:
     name: my-cluster
     namespace: default
 ---
```
Comment (sbueringer) [inline test/infrastructure/docker/examples/machine-pool.yaml:24]: fyi. I synced the examples so they are all very similar now (including ordering). I think that's the only way we can really maintain them over time

**Comment by fabriziopandini** [comment]
@chrischdi @sbueringer thanks for stepping up in fixing CAPD and our CI to adapt to recent changes in kind, really appreciated!
Let's make sure we got a green test grid before backporting.
/lgtm
/approve

**Comment by sbueringer** [comment]
> @chrischdi @sbueringer thanks for stepping up in fixing CAPD and our CI to adapt to recent changes in kind, really appreciated! Let's make sure we got a green test grid before backporting. /lgtm /approve

Thx! But the majority of the work was done by @chrischdi and @killianmuldoon, thx folks!!

I'll check testgrid etc. ... .
